airbyte-server/src/main/java/io/airbyte/server/handlers/helpers/CatalogConverter.java

```diff
@@ -30,7 +42,30 @@ private static io.airbyte.api.model.generated.AirbyteStream toApi(final io.airby
         .namespace(stream.getNamespace());
   }
 
-  private static io.airbyte.protocol.models.AirbyteStream toProtocol(final io.airbyte.api.model.generated.AirbyteStream stream) {
+  @SuppressWarnings("PMD.AvoidLiteralsInIfCondition")
+  private static io.airbyte.protocol.models.AirbyteStream toProtocol(final AirbyteStream stream, AirbyteStreamConfiguration config)
+      throws JsonValidationException {
+    JsonNode streamSchema = stream.getJsonSchema();
+    if (config.getFieldSelectionEnabled() != null && config.getFieldSelectionEnabled()) {
+      // Validate the selected field paths.
+      for (final var selectedFieldInfo : config.getSelectedFields()) {
+        if (selectedFieldInfo.getFieldPath().isEmpty()) {
+          throw new JsonValidationException("Selected field path cannot be empty");
+        }
+        if (selectedFieldInfo.getFieldPath().size() > 1) {
+          // TODO(mfsiega-airbyte): support nested fields.
+          throw new UnsupportedOperationException("Nested field selection not supported");
+        }
+      }
+      // Only include the selected fields.
+      List<String> selectedFieldNames = config.getSelectedFields().stream().map((field) -> field.getFieldPath().get(0)).collect(Collectors.toList());
+      final JsonNode properties = streamSchema.findValue("properties");
+      if (properties.isObject()) {
+        ((ObjectNode) properties).retain(selectedFieldNames);
+      } else {
+        throw new JsonValidationException("Requested field selection but no properties node found");
+      }
+    }
     return new io.airbyte.protocol.models.AirbyteStream()
         .withName(stream.getName())
         .withJsonSchema(stream.getJsonSchema())
```

Comment:
is this `stream.getJsonSchema()` a reference to the `streamSchema` variable that you've define and modified? Just making sure we're actually setting the updated schema value in the return value here.

Reply:
It's indeed modifying by reference (or by pointer, if you're a C++ user and a pedant) so it's all the same, but I think it isn't being used often enough to justify a local variable.